### PR TITLE
Unresolved models

### LINKED ISSUE #229: Improving model decorators
List of improvements for model decorators:

- [x] Possible to improve association decorator typing using `keyof` similar to Jasmine `spyOn` function where methods can be provided as a string representing the key:
  ```typescript
  @HasMany(SHALLOW_SITE, (m: Project) => m.siteIds, (m:Project) => m.secondKey, ...)
  // vs
  @HasMany(SHALLOW_SITE, "siteIds", "secondKey", ...)
  ```
  - This would somewhat simplify the Association Decorator code as the key can be used instead of the callback function
- [x] Possible to add typing to attribute decorator using a similar method to the one discussed above. Currently the key option is not type checked.
- [ ] ~Potentially standardise `@BawPersistAttr` decorator to follow the same system as other decorators. i.e.~
  ```typescript
  @BawAttribute({persist: true})
  ```
  - However this is based on the premise that `@BawAttribute` by itself would have some purpose
  - Could be an idea just to turn it into a function so that it is at least the same format as the other decorators
- [ ] Attempt to add more intelligent warnings, currently models will spam the console when an associated model's primary key is null or undefined
- [ ] Somehow remove the need for `InjectionTokens` (eg. `SHALLOW_SITE`), currently `InjectionTokens` are used as a workaround to prevent cyclic dependencies between models
- [x] Simplify associated model output.
  ```typescript
  projects.sites // observable
  // vs
  projects.sites // array
  ```
- [ ] Re-introduce access to associated model observable

### PR DESCRIPTION
**Pull request by Allcharles**
# Modify associated model typing

Removed wrapping observable for associated models. Change to models can be seen below:

```typescript
// before
projects.sites // Observable<Site[]>
// after
projects.sites // Array<Site>
```

## Changes

- Renamed `AudioRecordingComponent` from detail.ts to details.ts so it matches the naming convention of other files

## Issues

- Removed access to associated model observable, however it should take minimal effort in the future to re-add if required

## Closes

Related to #229

### CODE REVIEW

File: src/app/component/shared/detail-view/detail-view.component.spec.ts [hunk 1]
```diff
@@ -150,9 +150,36 @@ describe("DetailViewComponent", () => {
...
+        expect(true).toBeFalsy();
+      });
+
+      it("should handle hasOne unresolved associated model", () => {
+        expect(true).toBeFalsy();
+      });
+
+      it("should handle hasMany unresolved associated model", () => {
+        expect(true).toBeFalsy();
+      });
```
Comment (atruskie) [inline src/app/component/shared/detail-view/detail-view.component.spec.ts]: I assume these are intentionally failing?

File: src/app/component/shared/detail-view/render-field/render-field.component.ts [hunk 2]
```diff
@@ -43,13 +43,14 @@ import { toRelative } from "src/app/interfaces/apiInterfaces";
...
+        [field]="child"
       ></baw-render-field>
     </ng-template>
   `,
 })
-export class RenderFieldComponent extends WithUnsubscribe() implements OnInit {
-  @Input() view: ModelView;
+export class RenderFieldComponent extends WithUnsubscribe()
+  implements OnChanges {
+  @Input() field: ModelView;
```
Comment (atruskie) [inline src/app/component/shared/detail-view/render-field/render-field.component.ts]: Given the number of places where you had used `value` below I feel that's an indication to what the name of this property should be. Suggested name: `value`

File: src/app/models/AbstractModel.ts [hunk 3]
```diff
@@ -101,3 +103,39 @@ export abstract class AbstractModel {
...
+  private static readonly models = [];
+
+  static get one(): UnresolvedModel {
+    return this.model;
+  }
+
+  static get many(): AbstractModel[] {
+    return this.models;
+  }
+}
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts]: Merge these definitions into `UnresolvedModel`, remove `SingletonUnresolvedModel`

File: src/app/models/AbstractModel.ts [hunk 4]
```diff
@@ -101,3 +103,39 @@ export abstract class AbstractModel {
     return this[AbstractModel.metaKey];
   }
 }
+
+export class UnresolvedModel extends AbstractModel {
+  public readonly resolved = false;
+  public readonly kind = "UnresolvedModel";
+
+  constructor() {
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts:115]: add explicit accessibility modifier?

File: src/app/models/AbstractModel.ts [hunk 5]
```diff
@@ -101,3 +103,39 @@ export abstract class AbstractModel {
...
+
+  static get many(): AbstractModel[] {
+    return this.models;
+  }
+}
+
+export function isResolvedModel(
+  model: AbstractModel | UnresolvedModel
+): model is UnresolvedModel {
+  return model.resolved;
```
Comment (atruskie) [inline src/app/models/AbstractModel.ts]: I think we should just do a type check here and remove the `resolved` property completely.

File: src/app/models/AssociationDecorators.ts [hunk 6]
```diff
@@ -165,17 +164,22 @@ function createModelDecorator<M extends AbstractModel, S>(
...
     // Create service and request from API
     const service = injector.get(serviceToken.token);
-    const request = createRequest(service, identifier, parameters).pipe(
-      map((response) => {
-        // Change cached value to BehaviorSubject so that it will instantly return
-        updateCache(target, cachedModelKey, new BehaviorSubject(response));
-        return response;
-      })
+    createRequest(service, identifier, parameters).subscribe(
+      (model) => updateCache(target, cachedModelKey, model),
```
Comment (atruskie) [inline src/app/models/AssociationDecorators.ts]: Ok, this is great, however, if the observeable emits more than once we'll only get the last value.

Why might it omit more than once? Say we want the first two pages of API results to load by default for all associations.

Also, `updateCache` should probably be renamed to `updateBackingField`

**Review by atruskie — CHANGES_REQUESTED** [review]
Looks great! I was confused by how simple the changes ended up being. Nice work.

Minor changes needed for naming and esoteric corner case.

File: src/app/models/AbstractModel.ts (hunk 3 above)
Comment (Allcharles) [reply src/app/models/AbstractModel.ts]: Am i able to do that whilst also keeping the typings for AbstractModel the same as the AbstractModel?

Comment (atruskie) [reply src/app/models/AbstractModel.ts]: Should be able to? Nothing changes really. You can have static and non static members in one class... Therefore no reason to have a class with only static members in it.

Yes a static method can instantiate an instance of a type it is attached to.

File: src/app/models/AssociationDecorators.ts (hunk 6 above)
Comment (Allcharles) [reply src/app/models/AssociationDecorators.ts]: So if the response is in array, append to the array. What about single responses? Just override still?

Comment (atruskie) [reply src/app/models/AssociationDecorators.ts]: Yeah, override. I can't think of any other reasonable behaviour

**Review by atruskie — APPROVED** [review]
I'd suggest moving the abstract model tests from the detail view component. The tests are really about testing the AbstractModel association loading, and the detail view component is just a handy vehicle test bed.

The settimout fake style async is a little code-smelly. I've talked with you offline about it and I suggest dropping the interval to 0 where possible and also making a named function that makes the code easier to read. Something  like
```
await WaitOne()
```

**Comment by Allcharles** [comment]
> I'd suggest moving the abstract model tests from the detail view component. The tests are really about testing the AbstractModel association loading, and the detail view component is just a handy vehicle test bed.

So move the tests into the AssociatedDecorator test file?

**Comment by atruskie** [comment]
> > I'd suggest moving the abstract model tests from the detail view component. The tests are really about testing the AbstractModel association loading, and the detail view component is just a handy vehicle test bed.
> 
> So move the tests into the AssociatedDecorator test file?

For tests that are primarily related to the associations, yeah.

Another file is fine too if they're a different style of test. I'd be tempted to name it something like: `AssociationLoadingInComponents.spec.ts`

**Comment by Allcharles** [comment]
Okay, i wasn't a huge fan of extracting the tests as they were to `AssociationLoadingInComponents.spec.ts`, so i created a mock component so that future changes to the `DetailViewComponent` do not affect the tests. I have also kept the tests inside `DetailViewComponent` as they are still relevant to the component.

**Review by atruskie — APPROVED** [review]

**Pull request merged**